File: EZCropExample/EZCropExample/OC_Example/ViewController.m
//
//  ViewController.m
//  EZCropExample_OC
//
//  Created by Xiang Li on 4/8/21.
//

#import "ViewController.h"
@import EZCropController;

@interface ViewController () <UINavigationControllerDelegate, UIImagePickerControllerDelegate, EZCropControllerDelegate>

@property (nonatomic, strong) UIImage *image;           // The image we'll be cropping
@property (nonatomic, strong) UIImageView *imageView;   // The image view to present the cropped image

@property (nonatomic, assign) CGRect croppedFrame;
@property (nonatomic, assign) EZCropRotation angle;

@end

@implementation ViewController

#pragma mark - Image Picker Delegate -
- (void)imagePickerController:(UIImagePickerController *)picker didFinishPickingMediaWithInfo:(NSDictionary<NSString *,id> *)info
{
    UIImage *image = info[UIImagePickerControllerOriginalImage];
    self.image = image;

    [picker dismissViewControllerAnimated:YES completion:^{
        EZCropController *cropController = [[EZCropController alloc] initWithImage:image];
        cropController.delegate = self;
        [UIViewController swizzleShouldAutorotate];
        [self presentViewController:cropController animated:YES completion:nil];
    }];
}

- (void)imagePickerControllerDidCancel:(UIImagePickerController *)picker
{
    [picker dismissViewControllerAnimated:YES completion:nil];
}

#pragma mark - Gesture Recognizer -
- (void)didTapImageView
{
    // When tapping the image view, restore the image to the previous cropping state
    EZCropController* cropController = [[EZCropController alloc] initWithImage:self.image
                                                                      cropRect:self.croppedFrame
                                                                         angle:self.angle];
    cropController.delegate = self;
    [UIViewController swizzleShouldAutorotate];
    [self presentViewController:cropController animated:YES completion:nil];
}

#pragma mark - Cropper Delegate -

- (void)cropViewControllerCancel:(EZCropController * _Nonnull)cropViewController {
    [cropViewController dismissViewControllerAnimated:YES completion:^{
        [UIViewController swizzleShouldAutorotate];
    }];
}

- (void)cropViewController:(EZCropController *)cropViewController didCropTo:(UIImage *)image with:(CGRect)cropRect angle:(enum EZCropRotation)angle
{
    self.croppedFrame = cropRect;
    self.angle = angle;
    [self updateImageViewWithImage:image fromCropViewController:cropViewController];
}


- (void)updateImageViewWithImage:(UIImage *)image fromCropViewController:(EZCropController *)cropViewController
{
    self.imageView.image = image;
    [self layoutImageView];
    

    self.navigationItem.rightBarButtonItem.enabled = YES;

    self.imageView.hidden = NO;
    [cropViewController dismissViewControllerAnimated:YES completion:^{
        [UIViewController swizzleShouldAutorotate];
    }];
}

#pragma mark - Image Layout -
- (void)layoutImageView
{
    if (self.imageView.image == nil)
        return;

    CGFloat padding = 20.0f;

    CGRect viewFrame = self.view.bounds;
    viewFrame.size.width -= (padding * 2.0f);
    viewFrame.size.height -= ((padding * 2.0f));

    CGRect imageFrame = CGRectZero;
    imageFrame.size = self.imageView.image.size;

    if (self.imageView.image.size.width > viewFrame.size.width ||
        self.imageView.image.size.height > viewFrame.size.height)
    {
        CGFloat scale = MIN(viewFrame.size.width / imageFrame.size.width, viewFrame.size.height / imageFrame.size.height);
        imageFrame.size.width *= scale;
        imageFrame.size.height *= scale;
        imageFrame.origin.x = (CGRectGetWidth(self.view.bounds) - imageFrame.size.width) * 0.5f;
        imageFrame.origin.y = (CGRectGetHeight(self.view.bounds) - imageFrame.size.height) * 0.5f;
        self.imageView.frame = imageFrame;
    }
    else {
        self.imageView.frame = imageFrame;
        self.imageView.center = (CGPoint){CGRectGetMidX(self.view.bounds), CGRectGetMidY(self.view.bounds)};
    }
}

#pragma mark - Bar Button Items -
- (void)showCropViewController
{
    UIImagePickerController *standardPicker = [[UIImagePickerController alloc] init];
    standardPicker.sourceType = UIImagePickerControllerSourceTypePhotoLibrary;
    standardPicker.allowsEditing = NO;
    standardPicker.delegate = self;
    [self presentViewController:standardPicker animated:YES completion:nil];

}

- (void)sharePhoto:(id)sender
{
    if (self.imageView.image == nil)
        return;

    UIActivityViewController *activityController = [[UIActivityViewController alloc] initWithActivityItems:@[self.imageView.image] applicationActivities:nil];
    activityController.modalPresentationStyle = UIModalPresentationPopover;
    activityController.popoverPresentationController.barButtonItem = sender;
    [self presentViewController:activityController animated:YES completion:nil];
}

- (void)dismissViewController {
    [self dismissViewControllerAnimated:YES completion:nil];
}

#pragma mark - View Creation/Lifecycle -
- (void)viewDidLoad {
    [super viewDidLoad];
    self.title = NSLocalizedString(@"EZCropController", @"");

    self.navigationController.navigationBar.translucent = NO;

    self.navigationItem.leftBarButtonItem = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemAdd target:self action:@selector(showCropViewController)];

#if TARGET_APP_EXTENSION
    self.navigationItem.rightBarButtonItem = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemCancel target:self action:@selector(dismissViewController)];
#else
    self.navigationItem.rightBarButtonItem = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemAction target:self action:@selector(sharePhoto:)];
    self.navigationItem.rightBarButtonItem.enabled = NO;
#endif

    self.imageView = [[UIImageView alloc] init];
    self.imageView.userInteractionEnabled = YES;
    self.imageView.contentMode = UIViewContentModeScaleAspectFit;
    [self.view addSubview:self.imageView];

    if (@available(iOS 11.0, *)) {
        self.imageView.accessibilityIgnoresInvertColors = YES;
    }

    UITapGestureRecognizer *tapRecognizer = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(didTapImageView)];
    [self.imageView addGestureRecognizer:tapRecognizer];
}

- (void)viewDidLayoutSubviews
{
    [super viewDidLayoutSubviews];
    [self layoutImageView];
}





@end

File: EZCropExample/EZCropExample/swiftExample/ViewController.swift
//
//  ViewController.swift
//  EZCropExample
//
//  Created by Xiang Li on 4/8/21.
//

import UIKit
import EZCropController

class ViewController: UIViewController, UIImagePickerControllerDelegate, UINavigationControllerDelegate {



    private let imageView = UIImageView()

    private var image: UIImage?

    private var croppedRect = CGRect.zero
    private var croppedAngle : EZCropRotation = .zero
    /*public override var shouldAutorotate: Bool {
        if let presentedController = self.presentedViewController {
            return presentedController.shouldAutorotate
        }
        else {
            return true
        }
    }*/
    func imagePickerController(_ picker: UIImagePickerController, didFinishPickingMediaWithInfo info: [UIImagePickerController.InfoKey : Any]) {
        guard let image = (info[UIImagePickerController.InfoKey.originalImage] as? UIImage) else { return }
        self.shouldAutorotate
        let cropController =  EZCropController(image:image) //CropViewController(croppingStyle: croppingStyle, image: image)
        //cropController.modalPresentationStyle = .fullScreen
        cropController.delegate = self

        // Uncomment this if you wish to provide extra instructions via a title label
        //cropController.title = "Crop Image"

        // -- Uncomment these if you want to test out restoring to a previous crop setting --
        //cropController.angle = 90 // The initial angle in which the image will be rotated
        //cropController.imageCropFrame = CGRect(x: 0, y: 0, width: 2848, height: 4288) //The initial frame that the crop controller will have visible.

        // -- Uncomment the following lines of code to test out the aspect ratio features --
        //cropController.aspectRatioPreset = .presetSquare; //Set the initial aspect ratio as a square
        //cropController.aspectRatioLockEnabled = true // The crop box is locked to the aspect ratio and can't be resized away from it
        //cropController.resetAspectRatioEnabled = false // When tapping 'reset', the aspect ratio will NOT be reset back to default
        //cropController.aspectRatioPickerButtonHidden = true

        // -- Uncomment this line of code to place the toolbar at the top of the view controller --
        //cropController.toolbarPosition = .top

        //cropController.rotateButtonsHidden = true
        //cropController.rotateClockwiseButtonHidden = true

        //cropController.doneButtonTitle = "Title"
        //cropController.cancelButtonTitle = "Title"

        //cropController.toolbar.doneButtonHidden = true
        //cropController.toolbar.cancelButtonHidden = true
        //cropController.toolbar.clampButtonHidden = true

        // Set toolbar action button colors
        // cropController.doneButtonColor = UIColor.red
        // cropController.cancelButtonColor = UIColor.green

        self.image = image

        //If profile picture, push onto the same navigation stack
        picker.dismiss(animated: true, completion: {
            self.present(cropController, animated: true, completion: nil)
            //self.navigationController!.pushViewController(cropController, animated: true)
        })
    }

    /*public func cropViewController(_ cropViewController: CropViewController, didCropToImage image: UIImage, withRect cropRect: CGRect, angle: Int) {
        self.croppedRect = cropRect
        self.croppedAngle = angle
        updateImageViewWithImage(image, fromCropViewController: cropViewController)
    }

    public func cropViewController(_ cropViewController: CropViewController, didCropToCircularImage image: UIImage, withRect cropRect: CGRect, angle: Int) {
        self.croppedRect = cropRect
        self.croppedAngle = angle
        updateImageViewWithImage(image, fromCropViewController: cropViewController)
    }*/

    public func updateImageViewWithImage(_ image: UIImage, fromCropViewController cropViewController: EZCropController) {
        imageView.image = image
        layoutImageView()

        self.navigationItem.leftBarButtonItem?.isEnabled = true

        self.imageView.isHidden = false
        cropViewController.dismiss(animated: true, completion: nil)
    }

    override func viewDidLoad() {
        super.viewDidLoad()

        title = NSLocalizedString("CropViewController", comment: "")
        navigationController!.navigationBar.isTranslucent = false

        navigationItem.rightBarButtonItem = UIBarButtonItem(barButtonSystemItem: .add, target: self, action: #selector(addButtonTapped(sender:)))
        navigationItem.leftBarButtonItem = UIBarButtonItem(barButtonSystemItem: .action, target: self, action: #selector(sharePhoto))
        navigationItem.leftBarButtonItem?.isEnabled = false

        imageView.isUserInteractionEnabled = true
        imageView.contentMode = .scaleAspectFit
        if #available(iOS 11.0, *) {
            imageView.accessibilityIgnoresInvertColors = true
        }
        view.addSubview(imageView)

        let tapRecognizer = UITapGestureRecognizer.init(target: self, action: #selector(didTapImageView))
        imageView.addGestureRecognizer(tapRecognizer)
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }

    @objc public func addButtonTapped(sender: AnyObject) {
        let imagePicker = UIImagePickerController()
        imagePicker.sourceType = .photoLibrary
        imagePicker.allowsEditing = false
        imagePicker.delegate = self
        self.present(imagePicker, animated: true, completion: nil)
    }

    @objc public func didTapImageView() {
        // When tapping the image view, restore the image to the previous cropping state
        let cropViewController = EZCropController(image: self.image!, cropRect: self.croppedRect, angle: self.croppedAngle)
        cropViewController.delegate = self
        self.present(cropViewController, animated: true, completion: nil)
        //let cropViewController = CropViewController(croppingStyle: self.croppingStyle, image: self.image!)
        //cropViewController.delegate = self
       // let viewFrame = view.convert(imageView.frame, to: navigationController!.view)

        /*cropViewController.presentAnimatedFrom(self,
                                               fromImage: self.imageView.image,
                                               fromView: nil,
                                               fromFrame: viewFrame,
                                               angle: self.croppedAngle,
                                               toImageFrame: self.croppedRect,
                                               setup: { self.imageView.isHidden = true },
                                               completion: nil)*/
    }

    public override func viewDidLayoutSubviews() {
        super.viewDidLayoutSubviews()
        layoutImageView()
    }

    public func layoutImageView() {
        guard imageView.image != nil else { return }

        let padding: CGFloat = 20.0

        var viewFrame = self.view.bounds
        viewFrame.size.width -= (padding * 2.0)
        viewFrame.size.height -= ((padding * 2.0))

        var imageFrame = CGRect.zero
        imageFrame.size = imageView.image!.size;

        if imageView.image!.size.width > viewFrame.size.width || imageView.image!.size.height > viewFrame.size.height {
            let scale = min(viewFrame.size.width / imageFrame.size.width, viewFrame.size.height / imageFrame.size.height)
            imageFrame.size.width *= scale
            imageFrame.size.height *= scale
            imageFrame.origin.x = (self.view.bounds.size.width - imageFrame.size.width) * 0.5
            imageFrame.origin.y = (self.view.bounds.size.height - imageFrame.size.height) * 0.5
            imageView.frame = imageFrame
        }
        else {
            self.imageView.frame = imageFrame;
            self.imageView.center = CGPoint(x: self.view.bounds.midX, y: self.view.bounds.midY)
        }
    }

    @objc public func sharePhoto() {
        guard let image = imageView.image else {
            return
        }

        let activityController = UIActivityViewController(activityItems: [image], applicationActivities: nil)
        activityController.popoverPresentationController?.barButtonItem = navigationItem.rightBarButtonItem!
        present(activityController, animated: true, completion: nil)
    }
}

extension ViewController :  EZCropControllerDelegate{
    func cropViewControllerCancel(_ cropViewController: EZCropController) {
        cropViewController.dismiss(animated: true, completion: nil)
    }

    internal func cropViewController(_ cropViewController: EZCropController, didCropTo image: UIImage, with cropRect: CGRect, angle: EZCropRotation) {
        self.croppedRect = cropRect
        self.croppedAngle = angle
        updateImageViewWithImage(image, fromCropViewController: cropViewController)
    }
}

